### PR TITLE
fix: requeue if socket is destroying

### PR DESCRIPTION
**Pull request by ronag**

### CODE REVIEW

File: test/client-errors.js [hunk 1]
```diff
@@ -549,7 +549,9 @@ test('socket fail while ending request body', (t) => {
 
     const _err = new Error('kaboom')
     client.on('connect', () => {
-      client.socket.destroy(_err)
+      process.nextTick(() => {
```
Comment (mcollina) [inline test/client-errors.js:552]: why this nextTick?

Comment (ronag) [reply test/client-errors.js:552]: Because otherwise the socket will be destroyed before the queue has been resumed, which kind of invalidates the test.

**Comment by ronag** [comment]
I've missed something. Test failures are valid.

File: test/client-errors.js (hunk 1 above)
Comment (ronag) [reply test/client-errors.js:552]: changed to resume before emit.

**Comment by ronag** [comment]
valid failure on node < 13

**Comment by ronag** [comment]
Found some timing bugs that got activated with this changed. Was different between different version of node as well. Fixed.

File: lib/client.js [hunk 2]
```diff
@@ -440,24 +442,25 @@ class Client extends EventEmitter {
     this[kRetryTimeout] = null
 
     const onDestroyed = (err) => {
-      for (const callback of this[kDestroyed]) {
+      const callbacks = this[kDestroyed]
+      this[kDestroyed] = null
```
Comment (ronag) [inline lib/client.js:446]: There was a re-entrat bug here

File: lib/client.js [hunk 3]
```diff
@@ -440,24 +442,25 @@ class Client extends EventEmitter {
...
-          _err = err
-        })
         .on('close', () => {
-          onDestroyed(_err)
+          onDestroyed(this.socket[kError])
         })
         .destroy(err)
-    } else {
-      process.nextTick(onDestroyed, null, null)
     }
```
Comment (ronag) [inline lib/client.js:464]: Was a problem here when `client.close` was invoked between `'error'` and `'close'` being emitted on the socket.

File: test/client-errors.js [hunk 4]
```diff
@@ -519,7 +517,9 @@ test('socket fail while writing request body', (t) => {
     body.push('asd')
 
     client.on('connect', () => {
-      client.socket.destroy('kaboom')
+      process.nextTick(() => {
+        client.socket.destroy('kaboom')
+      })
```
Comment (ronag) [inline test/client-errors.js:522]: We resume the queue *after* emitting connect, thus we need to defer the destroy to keep the test valid.

**Pull request merged**